### PR TITLE
Lesson3

### PR DESCRIPTION
**Pull request by NataliMaevskaya**
**Task 3:**

**1)** Create a Multiview in the right part of the second row. Multiview should have 4 cells:
Ready-made Layout with Datatable and Form (Dashboard view)
Template for Users view
Template for Products view
Template for Admin view

**2)** Enable switching between Multiview cells using List on the left
Working with data:
Get data sources by the link: 
https://files.webix.com/30d/36d46ab8e511adfeda892d11b2da145f/data.zip

**3)**  First cell: Dashboard view (Datatable and form)
Load the Datatable data (“data/data.js”) by url
Enable data editing via Form. This short tutorial may help you: https://webix.com/quick-start/introduction.html#!/1
Enable deleting of datatable records via a “delete” icon in the last datatable column.
Add sorting and filtering to all Datatable columns (as in the picture below)
Make the first column look like the datatable header (same grey background)
Highlight rows when mouse pointer hovers over them

**4)** Second cell: Users view
Create List and Chart and load the same “data/users.js” to both of them
Enable List filtering via a Text input
Enable sorting via related buttons
Add “Delete” icons to all list items to remove the corresponding items
Highlight first 5 items (add some different background to them)
 
**5)** Third cell: Products view, hierarchy
Create a TreeTable with company products
Load “products.js”
Provide a template for displaying expand/collapse and folder icons in the “title” column
Enable cell selection
Open all branches after data is loaded

### CODE REVIEW

**Comment by NataliMaevskaya** [comment]
waiting for code review

File: script/script.js [hunk 1]
```diff
@@ -193,8 +344,36 @@ webix.ready(function () {
...
     });
+    $$("list").select("dashboard");
+
+    // filtering user's data
+    const usersList = $$("usersList");
+    $$("filterUsersList").attachEvent("onTimedKeypress", function () {
+        const filterText = this.getValue().toLowerCase();
+
+        usersList.filter(function (obj) {
+            const strForSearch = (obj.name + obj.country).toLowerCase();
```
Comment (Zwillinge) [inline script/script.js]: not a very good idea. you can find John from Belarus by a sub string "hnbe", which is not correct.

File: script/script.js [hunk 2]
```diff
@@ -136,44 +154,177 @@ const form = {
...
+    template: "#name# from #country# <span class='webix_icon wxi-close btn-right '></span>",
+    select: true,
+    css: "first-five-items",
+    onClick: {
+        "wxi-close": function (e, id) {
+            this.remove(id);
+            return false;
+        }
+    },
+    data: webix.copy(users)
```
Comment (Zwillinge) [inline script/script.js]: according to the task, you should load data by url. this is true for all data widgets in this task

File: script/script.js [hunk 3]
```diff
@@ -136,44 +154,177 @@ const form = {
...
+    scroll: "y",
+    select: true,
+    url: "../data/data.js",
+    hover: "hover-style",
+    sort: "multi",
+    columns: datatableColumns,
+    onClick: {
+        "wxi-trash": function (e, id) {
+            this.remove(id);
+            if ($$("form").getValues()) {
```
Comment (Zwillinge) [inline script/script.js]: here you clear the form anyway after deleting, but this makes sense only when the form contains the deleted record (you can compare ids)

**Review by Zwillinge — CHANGES_REQUESTED** [review]

File: data/dataProducts.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
+const products = [{
+		"id": "1",
+		"title": "Cinema",
+		"open":true,
```
Comment (Zwillinge) [inline data/dataProducts.js]: do not add `open:true` directly to the data. treat this data as if they came from backend you cannot modify.
you need to find a way to open the nodes after the data has been loaded into the widget

**Comment by NataliMaevskaya** [comment]
Corrections are done. Waiting for code review

File: script/script.js (hunk 1 above)
Comment (vadxb) [reply script/script.js]: please move all events into component's config

File: script/script.js [hunk 5]
```diff
@@ -193,8 +344,36 @@ webix.ready(function () {
...
+    });
+
+    // sorting users by name ascending
+    $$("sortAscBtn").attachEvent("onItemClick", function () {
+        usersList.sort("#name#", "asc", "string");
+        // usersList.sort("#age#", "asc", "int");
+    });
+
+    // sorting users by name descending
+    $$("sortDescBtn").attachEvent("onItemClick", function () {
```
Comment (vadxb) [inline script/script.js]: the same - move event inside of button config

**Review by vadxb — CHANGES_REQUESTED** [review]

File: script/script.js (hunk 5 above)
Comment (NataliMaevskaya) [reply script/script.js]: done

**Review by vadxb — APPROVED** [review]

File: script/script.js [hunk 6]
```diff
@@ -136,44 +154,205 @@ const form = {
...
+    scroll: "y",
+    select: true,
+    url: "../data/data.js",
+    hover: "hover-style",
+    sort: "multi",
+    columns: datatableColumns,
+    onClick: {
+        "wxi-trash": function (e, id) {
+            this.remove(id);
+            const itemIdNumber = Number(id);
```
Comment (Zwillinge) [inline script/script.js]: unnecessary type conversion, you can simply compare them as `id.row==values.id`.

**Review by Zwillinge — CHANGES_REQUESTED** [review]

File: script/script.js (hunk 6 above)
Comment (NataliMaevskaya) [reply script/script.js]: done

**Review by Zwillinge — APPROVED** [review]

**Pull request merged**